src/hotspot/share/opto/escape.cpp

```diff
@@ -3712,6 +3733,40 @@ void ConnectionGraph::dump(GrowableArray<PointsToNode*>& ptnodes_worklist) {
     }
   }
 }
+
+void ConnectionGraph::trace_es_update_helper(PointsToNode* ptn, PointsToNode::EscapeState es, bool fields, const char* reason) const {
+  if (_compile->directive()->TraceEscapeAnalysisOption) {
+    assert(ptn != nullptr, "should not be null");
+    assert(reason != nullptr, "should not be null");
+    ptn->dump_header(true);
+    PointsToNode::EscapeState new_es = fields ? ptn->escape_state() : es;
+    PointsToNode::EscapeState new_fields_es = fields ? es : ptn->fields_escape_state();
+    tty->print_cr("-> %s(%s) %s", esc_names[(int)new_es], esc_names[(int)new_fields_es], reason);
+  }
+}
+
+const char* ConnectionGraph::trace_propagate_message(PointsToNode* from) const {
+  if (_compile->directive()->TraceEscapeAnalysisOption) {
+    stringStream ss;
+    ss.print("propagated from: ");
+    from->dump(true, &ss, false);
+    return ss.as_string();
+  } else {
+    return nullptr;
+  }
+}
+
+const char* ConnectionGraph::trace_arg_escape_message(CallNode* call) const {
+  if (_compile->directive()->TraceEscapeAnalysisOption) {
+    stringStream ss;
+    ss.print("escapes as arg to:");
+    call->dump("", false, &ss);
+    return ss.as_string();
```

Comment:
stringStream::as_string will allocate a new string on resource arena. 
I think  you should place "ResourceMark rm;" to reclaim that space. 
ps: I don't quite understand why not all code of share/opto follow this idiom. maybe it's not an issue.

Reply:
There is already a `ResourceMark` in [`ConnectionGraph::do_analysis`](1) that catches these strings. FWIW adding another more nested `ResourceMark` is not possible since some of the GrowableArrays might then be resized in the nested mark which is a bug [(See)](2).

[1]: https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/escape.cpp#L96
[2]: https://github.com/openjdk/jdk/blob/master/src/hotspot/share/utilities/growableArray.cpp#L64-L71

Reply:
I miss that. you are right!